src/mrps.jl

```diff
@@ -0,0 +1,243 @@
+MRP{T} <: Rotation
+```
+Modified Rodrigues Parameter. Is a 3D parameterization of attitude, and is a sterographic
+projection of the 4D unit sphere onto the plane tangent to the negative real pole. They
+have a singularity at θ = ±180°.
+
+# Constructors
+MRP(x, y, z)
+MRP(r::SVector{3})
+"""
+struct MRP{T} <: Rotation{3,T}
+    x::T
+    y::T
+    z::T
+    @inline MRP{T}(x,y,z) where T = new{T}(x,y,z)
+end
+
+# ~~~~~~~~~~~~~~~ Constructors ~~~~~~~~~~~~~~~ #
+MRP(x::X, y::Y, z::Z) where {X,Y,Z} = MRP{promote_type(X,Y,Z)}(x, y, z)
+(::Type{M})(g::SVector{3}) where M<:MRP= M(g[1], g[2], g[3])
+
+# ~~~~~~~~~~~~~~~ Conversions ~~~~~~~~~~~~~~~ #
+SVector(g::MRP{T}) where T = SVector{3,T}(g.x, g.y, g.z)
+
+# ~~~~~~~~~~~~~~~ Initializers ~~~~~~~~~~~~~~~ #
+Base.rand(::Type{RP}) where RP <: MRP = RP(rand(UnitQuaternion))
+Base.one(::Type{RP}) where RP <: MRP = RP(0.0, 0.0, 0.0)
+
+# ~~~~~~~~~~~~~~~~ Quaternion <=> MRP ~~~~~~~~~~~~~~~~~~ #
+@inline function (::Type{Q})(p::MRP) where Q <: UnitQuaternion
+    n2 = p.x * p.x + p.y * p.y + p.z * p.z
+    M = 2/(1+n2)
+    q = Q((1-n2)/(1+n2), M*p.x, M*p.y, M*p.z, false)
+end
+
+@inline function (::Type{RP})(q::UnitQuaternion) where RP<:MRP
+    M = 1/(1+q.w)
+    RP(q.x*M, q.y*M, q.z*M)
+end
+
+# ~~~~~~~~~~~~~~~ StaticArrays Interface ~~~~~~~~~~~~~~~ #
+@inline (::Type{RP})(t::NTuple{9}) where RP<:MRP = convert(RP, UnitQuaternion(t))
+@inline Base.getindex(p::MRP, i::Int) = convert(UnitQuaternion, p)[i]
+@inline Base.Tuple(p::MRP) = Tuple(convert(UnitQuaternion, p))
+
+# ~~~~~~~~~~~~~~~ Math Operations ~~~~~~~~~~~~~~~ #
+LinearAlgebra.norm(p::MRP) = sqrt(p.x^2 + p.y^2 + p.z^2)
+
+
+# ~~~~~~~~~~~~~~~ Composition ~~~~~~~~~~~~~~~ #
+function (*)(p2::MRP, p1::MRP)
+    p2, p1 = SVector(p2), SVector(p1)
+    MRP(((1-p2'p2)*p1 + (1-p1'p1)*p2 - cross(2p1, p2) ) / (1+p1'p1*p2'p2 - 2p1'p2))
+end
+
+function (\)(p1::MRP, p2::MRP)
+    p1,p2 = SVector(p1), SVector(p2)
+    n1,n2 = p1'p1, p2'p2
+    θ = 1/((1+n1)*(1+n2))
+    s1,s2 = (1-n1), (1-n2)
+    v1 = -2p1
+    v2 =  2p2
+    s = s1*s2 - v1'v2
+    v = s1*v2 + s2*v1 + v1 × v2
+
+    M = θ/(1+θ*s)
+    return MRP(v*M)
+end
+
+function (/)(p1::MRP, p2::MRP)
+    n1,n2 = norm2(p1),   norm2(p2)
+    θ = 1/((1+n1)*(1+n2))
+    s1,s2 = (1-n1), (1-n2)
+    p1,p2 = SVector(p1), SVector(p2)
+    v1 =  2p1
+    v2 = -2p2
+    s = s1*s2 - v1'v2
+    v = s1*v2 + s2*v1 + v1 × v2
+
+    M = θ/(1+θ*s)
+    MRP(v[1]*M, v[2]*M, v[3]*M)
+
+end
+
+
+# ~~~~~~~~~~~~~~~ Rotation ~~~~~~~~~~~~~~~ #
+@inline (*)(p::MRP, r::SVector) = UnitQuaternion(p)*r
+@inline (\)(p::MRP, r::SVector) = UnitQuaternion(p)\r
+
+
+# ~~~~~~~~~~~~~~~ Kinematics ~~~~~~~~~~~~~~~ #
+function kinematics(p::MRP, ω)
+    p = SVector(p)
+    A = @SMatrix [
+        1 + p[1]^2 - p[2]^2 - p[3]^2  2(p[1]*p[2] - p[3])      2(p[1]*p[3] + p[2]);
+        2(p[2]*p[1] + p[3])            1-p[1]^2+p[2]^2-p[3]^2   2(p[2]*p[3] - p[1]);
+        2(p[3]*p[1] - p[2])            2(p[3]*p[2] + p[1])      1-p[1]^2-p[2]^2+p[3]^2]
```

Comment:
Is this essentially specialized code for `convert(::Type{RotMatrix}, ::MRP)`?

Reply:
It's related, but no. I added a comment here citing the equation and a brief explanation of the relationship to the rotation matrix. In summary, `A = B * (1 + p'p)` and `B*B = R` where `R` is the rotation matrix.

Reply:
Great, thanks.